File: middleware/ledger/pin.py
# The MIT License (MIT)
#
# Copyright (c) 2021 RSK Labs Ltd
#
# Permission is hereby granted, free of charge, to any person obtaining a copy of
# this software and associated documentation files (the "Software"), to deal in
# the Software without restriction, including without limitation the rights to
# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
# of the Software, and to permit persons to whom the Software is furnished to do
# so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import os
import random
import string
import logging


class PinError(RuntimeError):
    pass


class BasePin:
    PIN_LENGTH = 8
    POSSIBLE_CHARS = string.ascii_letters + string.digits
    ALPHA_CHARS = string.ascii_letters

    @classmethod
    def generate_pin(cls):
        random.seed()
        pin = None
        while pin is None or not cls.is_valid(pin.encode()):
            pin = ""
            for i in range(cls.PIN_LENGTH):
                pin += random.choice(cls.POSSIBLE_CHARS)
        return pin.encode()

    @classmethod
    def is_valid(cls, pin, any_pin=False):
        if type(pin) != bytes:
            return False

        if not all(map(lambda c: chr(c) in cls.POSSIBLE_CHARS, pin)):
            return False

        if any_pin:
            return True

        if len(pin) != cls.PIN_LENGTH:
            return False

        return any(map(lambda c: chr(c) in cls.ALPHA_CHARS, pin))


# Handles PIN initialization and loading
class FileBasedPin(BasePin):
    @classmethod
    def new(cls, path):
        # Generate a new random pin and save to the given path. Return the
        # generated pin
        with open(path, 'wb') as file:
            pin = cls.generate_pin()
            file.write(pin)
        return pin

    def __init__(self, path, default_pin, force_change=False):
        self.logger = logging.getLogger("pin")
        self._path = path

        # Check pin file existence
        try:
            self.logger.info("Checking whether pin file '%s' exists", path)
            pin_file_exists = os.path.isfile(path)
        except Exception as e:
            self._error("Error checking file existence: %s" % format(e))

        # Load pin
        if pin_file_exists:
            try:
                self.logger.info("Loading PIN from '%s'", path)
                with open(path, 'rb') as file:
                    self._pin = file.read().strip()
            except Exception as e:
                self._error("Error loading pin: %s" % format(e))
        else:
            self.logger.info("Using default PIN")
            self._pin = default_pin

        # Make sure pin is valid
        if not self.is_valid(self._pin):
            self._error("Invalid pin: %s" % self._pin)

        # Pin needs to be changed if no pin file found or a pin change is forced
        self._needs_change = force_change or not pin_file_exists
        self._changing = False

    # Returns the current pin
    def get_pin(self):
        return self._pin

    def get_new_pin(self):
        if not self._changing:
            return None

        return self._new_pin

    # Indicates whether the current pin needs changing
    def needs_change(self):
        return self._needs_change

    # Starts a pin change and generates a new pin
    def start_change(self):
        self.logger.info("Starting PIN change")

        if self._changing or not self._needs_change:
            return

        self._changing = True
        self._new_pin = self.generate_pin()

        self.logger.info("New PIN generated")

    # Finalizes a pin change by commiting the changes to disk
    def commit_change(self):
        self.logger.info("Commiting PIN change to disk")
        if not self._changing:
            return

        try:
            with open(self._path, "wb") as file:
                file.write(self._new_pin)
        except Exception as e:
            self._error("Error commiting: %s" % format(e))

        self._pin = self._new_pin
        self._new_pin = None
        self._changing = False
        self._needs_change = False

        self.logger.info("PIN change committed to disk")

    # Aborts a pin change
    def abort_change(self):
        self.logger.info("Aborting PIN change")

        if not self._changing:
            return

        self._new_pin = None
        self._changing = False

        self.logger.info("PIN change aborted")

    def _error(self, msg):
        self.logger.error(msg)
        raise PinError(msg)

File: ledger/src/ui/src/pin.c
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2021 RSK Labs Ltd
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

#include <string.h>

#include "apdu.h"
#include "os.h"
#include "ui_err.h"
#include "pin.h"

// Helper macros for pin validation
#define IS_IN_RANGE(c, begin, end) (((c) >= (begin)) && ((c) <= (end)))
#define IS_ALPHA(c) (IS_IN_RANGE(c, 'a', 'z') || IS_IN_RANGE(c, 'A', 'Z'))
#define IS_NUM(c) IS_IN_RANGE(c, '0', '9')
#define IS_ALPHANUM(c) (IS_ALPHA(c) || IS_NUM(c))

#define PIN_LENGTH 8
#define PIN_BUFFER_LENGTH (PIN_LENGTH + 2)
// Internal PIN buffer used for authenticated operations
unsigned char G_pin_buffer[PIN_BUFFER_LENGTH];
// Helper macros for pin manipulation when prepended length is used
#define GET_PIN() ((unsigned char *)(G_pin_buffer + 1))
#define GET_PIN_LENGTH() strlen((const char *)GET_PIN())

/*
 * Implements RSK PIN command.
 *
 * Receives one byte at a time and updates the pin context, adding a null byte
 * at the end.
 *
 * @arg[in] rx      number of received bytes from the Host
 * @ret             number of transmited bytes to the host
 */
unsigned int update_pin_buffer(volatile unsigned int rx) {
    // Should receive 1 byte per call
    if (APDU_DATA_SIZE(rx) != 1) {
        THROW(ERR_UI_PROT_INVALID);
    }

    unsigned char index = APDU_OP();
    if ((index >= 0) && (index <= PIN_LENGTH)) {
        G_pin_buffer[index] = APDU_AT(DATA);
        G_pin_buffer[index + 1] = 0;
    }

    return 3;
}

/*
 * Implements RSK NEW PIN command.
 *
 * Sets and checks the device pin.
 *
 * @ret number of transmited bytes to the host
 */
unsigned int set_pin() {
#ifndef DEBUG_BUILD
    if (!is_pin_valid()) {
        THROW(ERR_UI_INVALID_PIN);
    }
#endif
    // Set PIN
    os_perso_set_pin(0, GET_PIN(), GET_PIN_LENGTH());
    // check PIN
    os_global_pin_invalidate();
    unsigned char output_index = CMDPOS;
    SET_APDU_AT(output_index++, 2);
    SET_APDU_AT(output_index++,
                os_global_pin_check(GET_PIN(), GET_PIN_LENGTH()));
    return output_index;
}

/*
 * Validates that the pin curently saved to the internal buffer has exactly
 * PIN_LENGTH alphanumeric characters with at least one alphabetic character.
 *
 * @ret     true if pin is valid, false otherwise
 */
bool is_pin_valid() {
    // PIN_LENGTH is the only length accepted
    if (GET_PIN_LENGTH() != PIN_LENGTH) {
        return false;
    }
    // Check if PIN is alphanumeric
    bool hasAlpha = false;
    for (int i = 0; i < PIN_LENGTH; i++) {
        if (!IS_ALPHANUM(GET_PIN()[i])) {
            return false;
        }
        if (hasAlpha || IS_ALPHA(GET_PIN()[i])) {
            hasAlpha = true;
        }
    }

    return hasAlpha;
}

/*
 * Fills the internal pin buffer with zeroes
 */
void clear_pin() {
    explicit_bzero(G_pin_buffer, sizeof(G_pin_buffer));
}

/*
 * Uses the pin currently saved to the internal pin buffer to unlock the device
 *
 * @arg[in] prepended_length true if the internal buffer includes a prepended
 *                           length byte, false otherwise
 * @ret                      1 if pin validated successfully, 0 otherwise
 */
unsigned int unlock_with_pin(bool prepended_length) {
    if (prepended_length) {
        return os_global_pin_check(GET_PIN(), GET_PIN_LENGTH());
    } else {
        return os_global_pin_check(G_pin_buffer,
                                   strlen((const char *)G_pin_buffer));
    }
}

/*
 * Sets the pin currently saved to the internal pin buffer as the device's pin.
 * This function assumes the pin is saved with a prepended length byte.
 */
void set_device_pin() {
    os_perso_set_pin(0, GET_PIN(), GET_PIN_LENGTH());
}